File: csharp/src/Vista.SDK/Transport/DataChannel/DataChannel.cs
using System.Collections;
using System.Diagnostics.CodeAnalysis;
using System.Text.RegularExpressions;

namespace Vista.SDK.Transport.DataChannel;

public sealed record DataChannelListPackage
{
    public required Package Package { get; set; }

    public DataChannelList DataChannelList => Package.DataChannelList;
}

// Pacakge
public sealed record Package
{
    public required Header Header { get; set; }
    public required DataChannelList DataChannelList { get; set; }
}

// Header
public sealed record Header
{
    public required ShipId ShipId { get; set; }
    public required ConfigurationReference DataChannelListId { get; set; }
    public VersionInformation? VersionInformation { get; set; } = new();
    public required string? Author { get; set; }
    public DateTimeOffset? DateCreated { get; set; } = DateTimeOffset.UtcNow;
    public Dictionary<string, object>? CustomHeaders { get; set; }
}

public sealed record ConfigurationReference
{
    public required string Id { get; set; }
    public string? Version { get; set; }
    public required DateTimeOffset TimeStamp { get; set; }
}

public sealed record VersionInformation
{
    public static readonly string AnnexCNamingRule = "dnv";
    public static readonly string AnnexCNamingSchemeVersion = "v2";

    // Properties
    public string NamingRule { get; set; } = AnnexCNamingRule;
    public string NamingSchemeVersion { get; set; } = AnnexCNamingSchemeVersion;
    public string? ReferenceUrl { get; set; } = "https://docs.vista.dnv.com";
}

// DataChannelList
public sealed record DataChannelList() : ICollection<DataChannel>
{
    private List<DataChannel> dataChannels = new();
    private Dictionary<string, DataChannel> shortIdMap = new();
    private Dictionary<LocalId, DataChannel> localIdMap = new();

    public IReadOnlyList<DataChannel> DataChannels => dataChannels.AsReadOnly();

    public int Count => dataChannels.Count;

    public bool IsReadOnly => false;

    public bool TryGetByShortId(string shortId, [MaybeNullWhen(false)] out DataChannel dataChannel) =>
        shortIdMap.TryGetValue(shortId, out dataChannel);

    public bool TryGetByLocalId(LocalId localId, [MaybeNullWhen(false)] out DataChannel dataChannel) =>
        localIdMap.TryGetValue(localId, out dataChannel);

    public void Add(DataChannel dataChannel)
    {
        Add([dataChannel]);
    }

    public void Add(IEnumerable<DataChannel> dcs)
    {
        foreach (var dataChannel in dcs)
        {
            if (localIdMap.ContainsKey(dataChannel.DataChannelId.LocalId))
                throw new ArgumentException(
                    $"DataChannel with LocalId {dataChannel.DataChannelId.LocalId} already exists"
                );
            if (dataChannel.DataChannelId.ShortId is not null)
            {
                if (shortIdMap.ContainsKey(dataChannel.DataChannelId.ShortId))
                    throw new ArgumentException(
                        $"DataChannel with ShortId {dataChannel.DataChannelId.ShortId} already exists"
                    );
                shortIdMap.Add(dataChannel.DataChannelId.ShortId, dataChannel);
            }
            dataChannels.Add(dataChannel);
            localIdMap.Add(dataChannel.DataChannelId.LocalId, dataChannel);
        }
    }

    public void Clear()
    {
        dataChannels.Clear();
        shortIdMap.Clear();
        localIdMap.Clear();
    }

    public bool Contains(DataChannel item) => dataChannels.Contains(item);

    public void CopyTo(DataChannel[] array, int arrayIndex) => dataChannels.CopyTo(array, arrayIndex);

    public IEnumerator<DataChannel> GetEnumerator() => dataChannels.GetEnumerator();

    public bool Remove(DataChannel item)
    {
        if (!localIdMap.Remove(item.DataChannelId.LocalId))
            return false;
        if (item.DataChannelId.ShortId is not null && !shortIdMap.Remove(item.DataChannelId.ShortId))
            return false;
        return dataChannels.Remove(item);
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    public DataChannel this[string shortId] => shortIdMap[shortId];
    public DataChannel this[int index] => dataChannels[index];
    public DataChannel this[LocalId localId] => localIdMap[localId];
}

// DataChannel
public sealed record DataChannel
{
    private Property? property;

    public required DataChannelId DataChannelId { get; set; }
    public required Property Property
    {
        get => property ?? throw new InvalidOperationException("Property property not set");
        set
        {
            if (value.Validate() is ValidateResult.Invalid invalid)
                throw new ArgumentException($"Invalid property - Messages='[{string.Join(", ", invalid.Messages)}]'");
            property = value;
        }
    }
}

// DataChannelId
public sealed record DataChannelId
{
    public required LocalId LocalId { get; set; }
    public required string? ShortId { get; set; }
    public NameObject? NameObject { get; set; } = new();
}

public sealed record NameObject
{
    public static readonly string AnnexCNamingRule =
        $"/{VersionInformation.AnnexCNamingRule}-{VersionInformation.AnnexCNamingSchemeVersion}";
    public string NamingRule { get; set; } = AnnexCNamingRule;
    public Dictionary<string, object>? CustomNameObjects { get; set; }
}

// Property
public sealed class Property
{
    public required DataChannelType DataChannelType { get; set; }
    public required Format Format { get; set; }
    public required Range? Range { get; set; }
    public required Unit? Unit { get; set; }

    // TODO : Validate Quality coding 'In the case of “IEC 61162-STATUS”, “A” (Data valid) and “V” (Data invalid) are used for the data quality.'
    public string? QualityCoding { get; set; }

    // TODO : Validate - Priority level and criteria for classification shall be in accordance with IEC 62923-1:2018, 6.2.2.1
    public required string? AlertPriority { get; set; }
    public string? Name { get; set; }
    public string? Remarks { get; set; }

    public Dictionary<string, object>? CustomProperties { get; set; }

    public ValidateResult Validate()
    {
        var messages = new List<string>();

        if (Format.IsDecimal && Range is null)
            messages.Add("Range is required for Decimal format type");
        if (Format.IsDecimal && Unit is null)
            messages.Add("Unit is required for Decimal format type");
        if (DataChannelType.IsAlert && AlertPriority is null)
            messages.Add("AlertPriority is required for Alert DataChannelType");
        if (messages.Count > 0)
            return new ValidateResult.Invalid(messages.ToArray());
        return new ValidateResult.Ok();
    }
}

public sealed record DataChannelType
{
    private string? type;
    private double? updateCycle;
    private double? calculationPeriod;

    public required string Type
    {
        get => type ?? throw new InvalidOperationException("DataChannelType not set");
        set
        {
            var names = ISO19848.Instance.GetDataChannelTypeNames(ISO19848.LatestVersion);
            var result = names.Parse(value);
            if (result is not DataChannelTypeNames.ParseResult.Ok ok)
                throw new ArgumentException($"Invalid data channel type {type}");
            type = ok.TypeName.Type;
        }
    }
    public double? UpdateCycle
    {
        get => updateCycle;
        set
        {
            if (value < 0)
                throw new ArgumentException($"Invalid update cycle {value}. Should be positive");
            updateCycle = value;
        }
    }
    public double? CalculationPeriod
    {
        get => calculationPeriod;
        set
        {
            if (value < 0)
                throw new ArgumentException($"Invalid calculation period {value}. Should be positive");
            calculationPeriod = value;
        }
    }

    internal bool IsAlert => Type == "Alert";
}

public sealed record Format
{
    private string? type;
    private FormatDataType? dataType;

    public required string Type
    {
        get => type ?? throw new InvalidOperationException("Format type not set");
        set
        {
            var names = ISO19848.Instance.GetFormatDataTypes(ISO19848.LatestVersion);
            var result = names.Parse(value);
            if (result is not FormatDataTypes.ParseResult.Ok ok)
                throw new ArgumentException($"Invalid format type {type}");
            type = ok.TypeName.Type;
            dataType = ok.TypeName;
        }
    }
    public required Restriction? Restriction { get; set; } = null;

    internal FormatDataType DataType => dataType ?? throw new InvalidOperationException("Format type not set");
    internal bool IsDecimal => Type == "Decimal";

    public ValidateResult ValidateValue(string value)
    {
        if (DataType.Validate(value, out _) is ValidateResult.Invalid invalid)
            return invalid;
        if (Restriction is not null)
            return Restriction.ValidateValue(value, this);
        return new ValidateResult.Ok();
    }
}

public sealed record Restriction
{
    private uint? totalDigits;

    /// <summary>Defines a list of acceptable values.</summary>
    public IReadOnlyList<string>? Enumeration { get; set; }

    /// <summary>Specifies the maximum number of decimal places allowed. Shall be equal to or greater than zero.</summary>
    public uint? FractionDigits { get; set; }

    /// <summary>Specifies the exact number of characters or list items allowed. Shall be equal to or greater than zero.</summary>
    public uint? Length { get; set; }

    /// <summary>Specifies the upper bounds for numeric values (the value shall be less than this value).</summary>
    public double? MaxExclusive { get; set; }

    /// <summary>Specifies the upper bounds for numeric values (the value shall be less than or equal to this value).</summary>
    public double? MaxInclusive { get; set; }

    /// <summary>Specifies the maximum number of characters or list items allowed. Shall be equal to or greater than zero.</summary>
    public uint? MaxLength { get; set; }

    /// <summary>Specifies the lower bounds for numeric values (the value shall be greater than this value).</summary>
    public double? MinExclusive { get; set; }

    /// <summary>Specifies the lower bounds for numeric values (the value shall be greater than or equal to this value).</summary>
    public double? MinInclusive { get; set; }

    /// <summary>Specifies the minimum number of characters or list items allowed. Shall be equal to or greater than zero.</summary>
    public uint? MinLength { get; set; }

    /// <summary>Defines the exact sequence of characters that are acceptable.</summary>
    public string? Pattern { get; set; }

    /// <summary>Specifies the exact number of digits allowed. Shall be greater than zero.</summary>
    public uint? TotalDigits
    {
        get => totalDigits;
        set
        {
            if (value is not null && value <= 0)
                throw new ArgumentException("TotalDigits should be greater than zero");
            totalDigits = value;
        }
    }

    /// <summary>Specifies how white space (line feeds, tabs, spaces, and carriage returns) is handled.</summary>
    public WhiteSpace? WhiteSpace { get; set; }

    public ValidateResult ValidateValue(string value, Format format)
    {
        if (Enumeration is not null && !Enumeration.Contains(value, StringComparer.Ordinal))
            return new ValidateResult.Invalid([$"Value {value} is not in the enumeration"]);
        ValidateResult result = format
            .DataType
            .Match<ValidateResult>(
                value,
                dec =>
                {
                    if (FractionDigits is not null && CountDecimalPlaces(dec) > FractionDigits)
                        return new ValidateResult.Invalid(["Value has more decimal places than allowed"]);
                    if (ValidateNumber((double)dec) is ValidateResult.Invalid invalid)
                        return invalid;
                    return new ValidateResult.Ok();
                },
                i =>
                {
                    if (ValidateNumber(i) is ValidateResult.Invalid invalid)
                        return invalid;
                    if (TotalDigits is not null)
                    {
                        var numDigist = Math.Floor(Math.Log10(Math.Abs(i)) + 1) != TotalDigits;
                        return new ValidateResult.Invalid(
                            [$"Value {i} has {numDigist} digits but should be {TotalDigits}"]
                        );
                    }
                    return new ValidateResult.Ok();
                },
                b => new ValidateResult.Ok(),
                str =>
                {
                    var length = str.Length;
                    // check length
                    if (Length is not null && length != Length)
                        return new ValidateResult.Invalid([$"Value {str} has length {length} but should be {Length}"]);
                    // check max length
                    if (MaxLength is not null && length >= MaxLength)
                        return new ValidateResult.Invalid(
                            [$"Value {str} has length {length} but should be less than {MaxLength}"]
                        );
                    // check min length
                    if (MinLength is not null && length <= MinLength)
                        return new ValidateResult.Invalid(
                            [$"Value {str} has length {length} but should be greater than {MinLength}"]
                        );
                    // check pattern
                    if (Pattern is not null && !Regex.IsMatch(str, Pattern))
                        return new ValidateResult.Invalid([$"Value {str} does not match pattern {Pattern}"]);

                    return new ValidateResult.Ok();
                },
                date => new ValidateResult.Ok()
            );

        return result;
    }

    private ValidateResult ValidateNumber(double number)
    {
        // check max exclusive
        if (MaxExclusive is not null && number >= MaxExclusive)
            return new ValidateResult.Invalid([$"Value {number} is greater than or equal to {MaxExclusive}"]);
        // check max inclusive
        if (MaxInclusive is not null && number > MaxInclusive)
            return new ValidateResult.Invalid([$"Value {number} is greater than {MaxInclusive}"]);
        // check min exclusive
        if (MinExclusive is not null && number <= MinExclusive)
            return new ValidateResult.Invalid([$"Value {number} is less than or equal to {MinExclusive}"]);
        // check min inclusive
        if (MinInclusive is not null && number < MinInclusive)
            return new ValidateResult.Invalid([$"Value {number} is less than {MinInclusive}"]);
        return new ValidateResult.Ok();
    }

    private static decimal CountDecimalPlaces(decimal dec)
    {
        int[] bits = decimal.GetBits(dec);
        ulong lowInt = (uint)bits[0];
        ulong midInt = (uint)bits[1];
        int exponent = (bits[3] & 0x00FF0000) >> 16;
        int result = exponent;
        ulong lowDecimal = lowInt | (midInt << 32);
        while (result > 0 && (lowDecimal % 10) == 0)
        {
            result--;
            lowDecimal /= 10;
        }

        return result;
    }
}

public enum WhiteSpace
{
    Preserve = 0,
    Replace = 1,
    Collapse = 2,
}

public sealed record Range
{
    private double low = double.MinValue;
    private double high = double.MaxValue;

    public required double Low
    {
        get => low;
        set
        {
            if (value > High)
                throw new ArgumentException($"Low value {value} should be less than high value {High}");
            low = value;
        }
    }
    public required double High
    {
        get => high;
        set
        {
            if (value < Low)
                throw new ArgumentException($"High value {value} should be greater than low value {Low}");
            high = value;
        }
    }
}

// TODO : Validate Unit according to ISO 80000, IEC 80000 of Table 4
public sealed record Unit
{
    public required string UnitSymbol { get; set; }
    public string? QuantityName { get; set; }

    public Dictionary<string, object>? CustomElements { get; set; }
}

File: python/src/vista_sdk/transport/data_channel/data_channel.py
"""Data channel implementation for transport module."""

from __future__ import annotations

import math
import re
from collections.abc import Iterable, Iterator
from datetime import datetime, timezone
from decimal import Decimal
from enum import Enum
from typing import Any, TypeVar

from vista_sdk.local_id import LocalId
from vista_sdk.result import Invalid, Ok, ValidateResult
from vista_sdk.transport.iso19848 import FormatDataType
from vista_sdk.transport.ship_id import ShipId
from vista_sdk.transport.time_series_data.time_series_data import (
    ConfigurationReference as TimeSeriesConfigurationReference,
)
from vista_sdk.transport.value import AnyValue

T = TypeVar("T")


class DataChannelListPackage:
    """Package containing data channel list."""

    def __init__(self, package: Package) -> None:
        """Initialize with package."""
        self.package = package

    @property
    def data_channel_list(self) -> DataChannelList:
        """Get the data channel list."""
        return self.package.data_channel_list


class Package:
    """Package containing header and data channel list."""

    def __init__(self, header: Header, data_channel_list: DataChannelList) -> None:
        """Initialize package."""
        self.header = header
        self.data_channel_list = data_channel_list


class Header:
    """Header containing ship ID and configuration information."""

    def __init__(
        self,
        ship_id: ShipId,
        data_channel_list_id: ConfigurationReference,
        author: str | None,
        version_information: VersionInformation | None = None,
        date_created: datetime | None = None,
        custom_headers: dict[str, Any] | None = None,
    ) -> None:
        """Initialize header."""
        self.ship_id = ship_id
        self.data_channel_list_id = data_channel_list_id
        self.author = author
        self.version_information = version_information or VersionInformation()
        self.date_created = date_created or datetime.now(timezone.utc)
        self.custom_headers = custom_headers


class ConfigurationReference:
    """Reference to a configuration."""

    def __init__(
        self, config_id: str, timestamp: datetime, version: str | None = None
    ) -> None:
        """Initialize configuration reference."""
        self.id = config_id
        self.version = version
        self.timestamp = timestamp

    def as_time_series_reference(self) -> TimeSeriesConfigurationReference:
        """Convert to TimeSeriesData ConfigurationReference."""
        return TimeSeriesConfigurationReference(self.id, self.timestamp)


class VersionInformation:
    """Version information for naming scheme."""

    ANNEX_C_NAMING_RULE = "dnv"
    ANNEX_C_NAMING_SCHEME_VERSION = "v2"

    def __init__(
        self,
        naming_rule: str | None = None,
        naming_scheme_version: str | None = None,
        reference_url: str | None = None,
    ) -> None:
        """Initialize version information."""
        self.naming_rule = naming_rule or self.ANNEX_C_NAMING_RULE
        self.naming_scheme_version = (
            naming_scheme_version or self.ANNEX_C_NAMING_SCHEME_VERSION
        )
        self.reference_url = reference_url or "https://docs.vista.dnv.com"


class DataChannelList:
    """Collection of data channels with lookup capabilities."""

    def __init__(self) -> None:
        """Initialize empty data channel list."""
        self._data_channels: list[DataChannel] = []
        self._short_id_map: dict[str, DataChannel] = {}
        self._local_id_map: dict[LocalId, DataChannel] = {}

    @property
    def data_channels(self) -> list[DataChannel]:
        """Get read-only list of data channels."""
        return self._data_channels.copy()

    def __len__(self) -> int:
        """Get number of data channels."""
        return len(self._data_channels)

    def try_get_by_short_id(self, short_id: str) -> tuple[bool, DataChannel | None]:
        """Try to get data channel by short ID."""
        data_channel = self._short_id_map.get(short_id)
        return data_channel is not None, data_channel

    def try_get_by_local_id(self, local_id: LocalId) -> tuple[bool, DataChannel | None]:
        """Try to get data channel by local ID."""
        data_channel = self._local_id_map.get(local_id)
        return data_channel is not None, data_channel

    def add(self, data_channel: DataChannel | Iterable[DataChannel]) -> None:
        """Add data channel(s) to the list."""
        if isinstance(data_channel, DataChannel):
            self._add_single(data_channel)
        else:
            self._add_multiple(data_channel)

    def _add_single(self, data_channel: DataChannel) -> None:
        """Add a single data channel."""
        local_id = data_channel.data_channel_id.local_id
        if local_id in self._local_id_map:
            raise ValueError(f"DataChannel with LocalId {local_id} already exists")

        short_id = data_channel.data_channel_id.short_id
        if short_id is not None:
            if short_id in self._short_id_map:
                raise ValueError(f"DataChannel with ShortId {short_id} already exists")
            self._short_id_map[short_id] = data_channel

        self._data_channels.append(data_channel)
        self._local_id_map[local_id] = data_channel

    def _add_multiple(self, data_channels: Iterable[DataChannel]) -> None:
        """Add multiple data channels."""
        for data_channel in data_channels:
            self._add_single(data_channel)

    def clear(self) -> None:
        """Clear all data channels."""
        self._data_channels.clear()
        self._short_id_map.clear()
        self._local_id_map.clear()

    def __contains__(self, item: DataChannel) -> bool:
        """Check if data channel is in the list."""
        return item in self._data_channels

    def remove(self, item: DataChannel) -> bool:
        """Remove data channel from the list."""
        local_id = item.data_channel_id.local_id
        if local_id not in self._local_id_map:
            return False

        short_id = item.data_channel_id.short_id
        if short_id is not None and short_id not in self._short_id_map:
            return False

        del self._local_id_map[local_id]
        if short_id is not None:
            del self._short_id_map[short_id]

        try:
            self._data_channels.remove(item)
            return True
        except ValueError:
            return False

    def __getitem__(self, key: str | int | LocalId) -> DataChannel:
        """Get data channel by key."""
        if isinstance(key, str):
            if key not in self._short_id_map:
                raise KeyError(f"No data channel with short ID '{key}'")
            return self._short_id_map[key]
        if isinstance(key, int):
            return self._data_channels[key]
        if isinstance(key, LocalId):
            if key not in self._local_id_map:
                raise KeyError(f"No data channel with local ID '{key}'")
            return self._local_id_map[key]
        raise TypeError(f"Invalid key type: {type(key)}")

    def __iter__(self) -> Iterator[DataChannel]:
        """Iterate over data channels."""
        return iter(self._data_channels)


class DataChannel:
    """Represents a data channel with ID and property."""

    def __init__(self, data_channel_id: DataChannelId, property_: Property) -> None:
        """Initialize data channel."""
        self.data_channel_id = data_channel_id
        self._property = None
        self.property_ = property_  # This will trigger validation

    @property
    def property_(self) -> Property:
        """Get the property."""
        if self._property is None:
            raise ValueError("Property not set")
        return self._property

    @property_.setter
    def property_(self, value: Property) -> None:
        """Set the property with validation."""
        validation_result = value.validate()
        if isinstance(validation_result, Invalid):
            messages = ", ".join(validation_result.messages)
            raise ValueError(f"Invalid property - Messages='[{messages}]'")
        self._property = value


class DataChannelId:
    """Data channel identifier with local ID and optional short ID."""

    def __init__(
        self,
        local_id: LocalId,
        short_id: str | None,
        name_object: NameObject | None = None,
    ) -> None:
        """Initialize data channel ID."""
        self.local_id = local_id
        self.short_id = short_id
        self.name_object = name_object or NameObject()


class NameObject:
    """Name object for data channel."""

    ANNEX_C_NAMING_RULE = (
        f"/{VersionInformation.ANNEX_C_NAMING_RULE}-"
        f"{VersionInformation.ANNEX_C_NAMING_SCHEME_VERSION}"
    )

    def __init__(
        self,
        naming_rule: str | None = None,
        custom_name_objects: dict[str, Any] | None = None,
    ) -> None:
        """Initialize name object."""
        self.naming_rule = naming_rule or self.ANNEX_C_NAMING_RULE
        self.custom_name_objects = custom_name_objects


class Property:
    """Property containing data channel type, format, range, and unit."""

    def __init__(
        self,
        data_channel_type: DataChannelType,
        data_format: Format,
        data_range: Range | None,
        unit: Unit | None,
        quality_coding: str | None = None,
        alert_priority: str | None = None,
        name: str | None = None,
        remarks: str | None = None,
        custom_properties: dict[str, Any] | None = None,
    ) -> None:
        """Initialize property."""
        self.data_channel_type = data_channel_type
        self.format = data_format
        self.range = data_range
        self.unit = unit
        self.quality_coding = quality_coding
        self.alert_priority = alert_priority
        self.name = name
        self.remarks = remarks
        self.custom_properties = custom_properties

    def validate(self) -> ValidateResult:
        """Validate the property."""
        messages = []

        if self.format.is_decimal and self.range is None:
            messages.append("Range is required for Decimal format type")
        if self.format.is_decimal and self.unit is None:
            messages.append("Unit is required for Decimal format type")
        if self.data_channel_type.is_alert and self.alert_priority is None:
            messages.append("AlertPriority is required for Alert DataChannelType")

        if messages:
            return Invalid(messages)
        return Ok()


class DataChannelType:
    """Data channel type with validation."""

    def __init__(
        self,
        channel_type: str,
        update_cycle: float | None = None,
        calculation_period: float | None = None,
    ) -> None:
        """Initialize data channel type."""
        self._type: str | None = None
        self._update_cycle: float | None = None
        self._calculation_period: float | None = None

        self.type = channel_type
        if update_cycle is not None:
            self.update_cycle = update_cycle
        if calculation_period is not None:
            self.calculation_period = calculation_period

    @property
    def type(self) -> str:
        """Get the data channel type."""
        if self._type is None:
            raise ValueError("DataChannelType not set")
        return self._type

    @type.setter
    def type(self, value: str) -> None:
        """Set the data channel type with validation."""
        # Note: In the C# version, this validates against ISO19848
        # data channel type names
        # For now, we'll do basic validation and can enhance later
        if not value:
            raise ValueError(f"Invalid data channel type {value}")
        self._type = value

    @property
    def update_cycle(self) -> float | None:
        """Get the update cycle."""
        return self._update_cycle

    @update_cycle.setter
    def update_cycle(self, value: float | None) -> None:
        """Set the update cycle with validation."""
        if value is not None and value < 0:
            raise ValueError(f"Invalid update cycle {value}. Should be positive")
        self._update_cycle = value

    @property
    def calculation_period(self) -> float | None:
        """Get the calculation period."""
        return self._calculation_period

    @calculation_period.setter
    def calculation_period(self, value: float | None) -> None:
        """Set the calculation period with validation."""
        if value is not None and value < 0:
            raise ValueError(f"Invalid calculation period {value}. Should be positive")
        self._calculation_period = value

    @property
    def is_alert(self) -> bool:
        """Check if this is an alert type."""
        return self.type == "Alert"


class Format:
    """Format specification with data type and restrictions."""

    def __init__(
        self, format_type: str, restriction: Restriction | None = None
    ) -> None:
        """Initialize format."""
        self._type: str | None = None
        self._data_type = None
        self.restriction = restriction
        self.type = format_type

    @property
    def type(self) -> str:
        """Get the format type."""
        if self._type is None:
            raise ValueError("Format type not set")
        return self._type

    @type.setter
    def type(self, value: str) -> None:
        """Set the format type with validation."""
        # Note: In the C# version, this validates against ISO19848 format data types
        # For now, we'll do basic validation and can enhance later
        if not value:
            raise ValueError(f"Invalid format type {value}")
        self._type = value
        # For now, we'll create a simple data type object
        self._data_type = FormatDataType(value)

    @property
    def data_type(self) -> FormatDataType:
        """Get the data type."""
        if self._data_type is None:
            raise ValueError("Format type not set")
        return self._data_type

    @property
    def is_decimal(self) -> bool:
        """Check if this is a decimal format."""
        return self.type == "Decimal"

    def validate_value(self, value: str) -> tuple[ValidateResult, AnyValue | None]:
        """Validate a value against this format."""
        result, parsed_value = self.data_type.validate(value)
        if isinstance(result, Invalid):
            return result, None

        if self.restriction is not None:
            restriction_result = self.restriction.validate_value(value, self)
            if isinstance(restriction_result, Invalid):
                return restriction_result, None

        return Ok(), parsed_value


class Restriction:
    """Validation restrictions for format values."""

    def __init__(
        self,
        enumeration: list[str] | None = None,
        fraction_digits: int | None = None,
        length: int | None = None,
        max_exclusive: float | None = None,
        max_inclusive: float | None = None,
        max_length: int | None = None,
        min_exclusive: float | None = None,
        min_inclusive: float | None = None,
        min_length: int | None = None,
        pattern: str | None = None,
        total_digits: int | None = None,
        white_space: WhiteSpace | None = None,
    ) -> None:
        """Initialize restriction."""
        self.enumeration = enumeration
        self.fraction_digits = fraction_digits
        self.length = length
        self.max_exclusive = max_exclusive
        self.max_inclusive = max_inclusive
        self.max_length = max_length
        self.min_exclusive = min_exclusive
        self.min_inclusive = min_inclusive
        self.min_length = min_length
        self.pattern = pattern
        self.white_space = white_space
        self._total_digits = None
        if total_digits is not None:
            self.total_digits = total_digits

    @property
    def total_digits(self) -> int | None:
        """Get total digits constraint."""
        return self._total_digits

    @total_digits.setter
    def total_digits(self, value: int | None) -> None:
        """Set total digits constraint with validation."""
        if value is not None and value <= 0:
            raise ValueError("TotalDigits should be greater than zero")
        self._total_digits = value

    def validate_value(self, value: str, format_obj: Format) -> ValidateResult:
        """Validate a value against the restrictions."""
        if self.enumeration is not None and value not in self.enumeration:
            return Invalid([f"Value {value} is not in the enumeration"])

        # Use match pattern to dispatch validation based on data type
        return format_obj.data_type.match(
            value,
            on_decimal=lambda dec: self._validate_decimal_value(dec),
            on_integer=lambda i: self._validate_integer_value(i),
            on_boolean=lambda _: Ok(),
            on_string=lambda s: self._validate_string(s),
            on_datetime=lambda _: Ok(),
        )

    def _validate_decimal_value(self, dec: Decimal) -> ValidateResult:
        """Validate decimal value against restrictions."""
        if self.fraction_digits is not None:
            decimal_places = self._count_decimal_places(str(dec))
            if decimal_places > self.fraction_digits:
                return Invalid(["Value has more decimal places than allowed"])
        number_result = self._validate_number(float(dec))
        if isinstance(number_result, Invalid):
            return number_result
        if self.total_digits is not None:
            num_digits = self._count_total_digits(dec)
            if num_digits != self.total_digits:
                return Invalid(
                    [
                        f"Value {dec} has {num_digits} digits "
                        f"but should be {self.total_digits}"
                    ]
                )
        return Ok()

    def _validate_integer_value(self, i: int) -> ValidateResult:
        """Validate integer value against restrictions."""
        validation_result = self._validate_number(float(i))
        if isinstance(validation_result, Invalid):
            return validation_result
        if self.total_digits is not None:
            num_digits = 1 if i == 0 else math.floor(math.log10(abs(i)) + 1)
            if num_digits != self.total_digits:
                return Invalid(
                    [
                        f"Value {i} has {num_digits} digits "
                        f"but should be {self.total_digits}"
                    ]
                )
        return Ok()

    def _validate_string(self, value: str) -> ValidateResult:
        """Validate string value against restrictions."""
        length = len(value)
        if self.length is not None and length != self.length:
            return Invalid(
                [f"Value {value} has length {length} but should be {self.length}"]
            )
        if self.max_length is not None and length > self.max_length:
            return Invalid(
                [
                    f"Value {value} has length {length} "
                    f"but should be at most {self.max_length}"
                ]
            )
        if self.min_length is not None and length < self.min_length:
            return Invalid(
                [
                    f"Value {value} has length {length} "
                    f"but should be at least {self.min_length}"
                ]
            )
        if self.pattern is not None and not re.match(self.pattern, value):
            return Invalid([f"Value {value} does not match pattern {self.pattern}"])

        return Ok()

    def _validate_number(self, number: float) -> ValidateResult:
        """Validate numeric constraints."""
        if self.max_exclusive is not None and number >= self.max_exclusive:
            return Invalid(
                [f"Value {number} is greater than or equal to {self.max_exclusive}"]
            )
        if self.max_inclusive is not None and number > self.max_inclusive:
            return Invalid([f"Value {number} is greater than {self.max_inclusive}"])
        if self.min_exclusive is not None and number <= self.min_exclusive:
            return Invalid(
                [f"Value {number} is less than or equal to {self.min_exclusive}"]
            )
        if self.min_inclusive is not None and number < self.min_inclusive:
            return Invalid([f"Value {number} is less than {self.min_inclusive}"])
        return Ok()

    def _count_decimal_places(self, value: str) -> int:
        """Count decimal places in a string representation of a number."""
        if "." not in value:
            return 0
        return len(value.split(".")[1])

    def _count_total_digits(self, value: float) -> int:
        """Count the total number of significant digits in a numeric value.

        Mirrors XSD totalDigits semantics: leading zeros and trailing
        fractional zeros are not counted, trailing integer zeros are.
        """
        s = repr(abs(value))
        if "e" in s or "E" in s:
            s = f"{abs(value):.20f}"
        if "." in s:
            s = s.rstrip("0").rstrip(".")
        s = s.replace(".", "").lstrip("0")
        return len(s) if s else 1


class WhiteSpace(Enum):
    """White space handling enumeration."""

    Preserve = 0
    Replace = 1
    Collapse = 2


class Range:
    """Numeric range specification."""

    def __init__(self, low: float, high: float) -> None:
        """Initialize range."""
        if low > high:
            raise ValueError(f"Low value {low} should be less than high value {high}")
        self._low = low
        self._high = high

    @property
    def low(self) -> float:
        """Get the low value."""
        return self._low

    @low.setter
    def low(self, value: float) -> None:
        """Set the low value with validation."""
        if value > self._high:
            raise ValueError(
                f"Low value {value} should be less than high value {self._high}"
            )
        self._low = value

    @property
    def high(self) -> float:
        """Get the high value."""
        return self._high

    @high.setter
    def high(self, value: float) -> None:
        """Set the high value with validation."""
        if value < self._low:
            raise ValueError(
                f"High value {value} should be greater than low value {self._low}"
            )
        self._high = value


class Unit:
    """Unit specification."""

    def __init__(
        self,
        unit_symbol: str,
        quantity_name: str | None = None,
        custom_elements: dict[str, Any] | None = None,
    ) -> None:
        """Initialize unit."""
        self.unit_symbol = unit_symbol
        self.quantity_name = quantity_name
        self.custom_elements = custom_elements
